src/compiler/crystal/command.cr

```diff
@@ -521,17 +520,29 @@ class Crystal::Command
     if has_stdin_filename
       sources << Compiler::Source.new(filenames.shift, STDIN.gets_to_end)
     end
-    sources += gather_sources(filenames)
-    first_filename = sources.first.filename
-    first_file_ext = File.extname(first_filename)
-    original_output_filename = File.basename(first_filename, first_file_ext)
+    sources.concat gather_sources(filenames)
 
-    # Check if we'll overwrite the main source file
-    if first_file_ext.empty? && !output_filename && !no_codegen && !run && first_filename == File.expand_path(original_output_filename)
-      error "compilation will overwrite source file '#{Crystal.relative_filename(first_filename)}', either change its extension to '.cr' or specify an output file with '-o'"
+    if output_filename
+      if compiler.codegen_target.win32?
+        # foo -> foo.exe
+        # foo.bar -> foo.exe
+        # foo.exe -> foo.exe
+        # foo.tar.gz -> foo.tar.exe
+        output_filename = "#{output_filename.rchop(File.extname(output_filename))}.exe"
```

Comment:
`output_filename` is what the user passed to the `-o` flag. So I'm doubting this behaviour is good. This flag expresses an explicit choice. I don't think the compiler should mess with that.

It might be reasonable to append `.exe` if missing for compliance. But IMO no part of the option value should be removed.